### ISSUE TITLE
Apple Silicon support

### ISSUE THREAD

**Issue by r58Playz** [MultiMC/Launcher#3480]
Are there any instructions for building on Apple Silicon? Are there any binaries?

**Comment by kb-1000** [comment]
No.
Ask again at Mojang, they have to distribute ARM natives for the game itself for this to even be considered.

**Comment by r58Playz** [comment]
Can you use the natives from [this](https://gist.github.com/tanmayb123/d55b16c493326945385e815453de411a)?

**Comment by r58Playz** [comment]
I got both MC 1.16 using LWJGL 3 and LWJGL 2 to run using Rosetta, though LWJGL 3 needed natives compiled for Apple M1.

**Comment by panther03** [comment]
> I got both MC 1.16 using LWJGL 3 to run using Rosetta, though LWJGL 3 needed natives compiled for Apple M1. Here is a link to the instance zip: https://drive.google.com/file/d/1cy2xHoik6psLMKWZmMw-pJpcYkp3ssM3/view?usp=sharing .

I just downloaded and launched your instance and I'm not sure if what you did for the libraries worked at all, or if it really has any effect. As far as I'm aware MultiMC doesn't search for natives in the instances libraries folder, and if it does, these libraries are not being used. I know this because when the game is launched with an x86_64 JVM it just runs in Rosetta, and if you launch it with an aarch64 JVM it complains about a JNI architecture mismatch (since it can't load dynamic libraries on a mismatching architecture).

Would love to know if there's a way to properly specify custom natives on MultiMC, since I can't get it to work. -Djava.library.path doesn't work either

**Comment by panther03** [comment]
Well, I'm not exactly proud of this, but if you really want to get MultiMC working natively on your apple silicon mac, currently only working with vanilla 1.13 and up (lwjgl3), here's what I did and know works:
- Download Qt5, compile from source (Can be from git, but it should be the Qt5.12 branch). Make sure it's targeting arm64, by default it will target x86_64
- Use the compiled qt5 to build MultiMC (instructions on the wiki still apply)
- Remove all of the extraction steps in ExtractNatives.cpp, not going to outline the exact lines of code but it should be pretty easy to figure out. Just want to stop multimc from extracting its own natives, and promptly removing the directory.
- change line 64 of LibrariesTask.cpp from: `connect(downloadJob.get(), &NetJob::failed, this, &LibrariesTask::jarlibFailed);` to `connect(downloadJob.get(), &NetJob::failed, this, &LibrariesTask::emitSucceeded);`
- make && make install your multimc source, then attempt to launch an instance for the first time. this should fail
- browse the package contents of your installed MultiMC.app, and do the following:
     - in each lwjgl submodule directory (libraries/org/lwjgl/lwjgl-<some-module>), delete the non-native jar
     - in the  libraries/org/lwjgl/lwjgl/3.2.2 directory, take lwjglfat.jar from the zip of the gist linked above and paste it here, renaming it to lwjgl-3.2.2.jar, which was deleted in the previous step
     - set the entire tree of org/lwjgl to read only, preventing multiMC from modifying it
- for each new instance you want to run, you will need to make the natives directory in that instance manually, and paste the natives from the gist above.
that should be it, make sure of course that you're using an arm64 JVM, like the ones available from zulu's website. this is obviously a huge hack and it would be better if you could supply custom manifests, which would then just point to your own library asset URLs. I tried just replacing the manifest that was already there, but multimc re-downloads it on every launch and I think it keeps the URLs cached in memory even if you replace it while it's running. Of course it would also be nice to have these lwjgl natives as part of the automatic builds on LWJGL/Mojang's end, but I guess we're not quite there yet.

I do think this is pretty important to have even if Rosetta exists, because the performance uplift is enormous. Also, I've heard differing experiences but newer minecraft doesn't run too well in Rosetta for me.

**Comment by peterix** [comment]
No.

And don't expect there to be any until Mojang ships arm64 LWJGL.

Meanwhile, it just works thanks to rosetta2.

I'm ready to do this and have a M1 mac to build/test set up... Just waiting on dependencies to materialize.

There is also no need to track this.

**Comment by nikhiljha** [comment]
This issue shows up first in google, followed by a repository with random compiled binaries 😟, so as a PSA: please try to avoid using binaries from unofficial sources!

If you want to run MultiMC on Apple Silicon with a native version of Java, I made a gist with instructions for how to use the binaries from the LWJGL website.

https://gist.github.com/nikhiljha/7313ac5553aafb1c8596b1fca0f4cdff

**Comment by tactlessfish** [comment]
I noticed that MultiMC works on Apple Silicon if you use a patched LWJGL 3 json [as per this issue](https://github.com/yusefnapora/m1-multimc-hack/issues/32). I don't completely understand how it works, would it be practical if MultiMC started using the patched json when creating new instances on Apple Silicon? @peterix

**Comment by QianNangong** [comment]
@r58Playz Hi, 1.19 will support apple silicon since there are entries referred to `natives-macos-arm64`
<img width="1324" alt="image" src="https://user-images.githubusercontent.com/4312700/167598180-74678fdc-4efe-4a52-b183-4ac2c1b88a29.png">
Will there be some update here?

**Comment by charlie-niekirk** [comment]
This is now officially supported by Mojang, any idea when we can get an easy way to create an ARM64 instance?

**Comment by johntdavis84** [comment]
Thanks for pointing that out, @charlie-niekirk . :)

I didn't realize official Apple Silicon support was so recent for Minecraft.

Is there a way to volunteer to help test?

**Comment by djfurman** [comment]
I have an M1 and trying to use MultiMC for the first time so I'm happy to test. I'm certainly failing now :)

**Comment by 1oglop1** [comment]
> @r58Playz Hi, 1.19 will support apple silicon since there are entries referred to `natives-macos-arm64` <img alt="image" width="1324" src="https://user-images.githubusercontent.com/4312700/167598180-74678fdc-4efe-4a52-b183-4ac2c1b88a29.png"> Will there be some update here?

@QianNangong It would be better if you also share the original URL as well

**Comment by QianNangong** [comment]
> > @r58Playz Hi, 1.19 will support apple silicon since there are entries referred to `natives-macos-arm64` <img alt="image" width="1324" src="https://user-images.githubusercontent.com/4312700/167598180-74678fdc-4efe-4a52-b183-4ac2c1b88a29.png"> Will there be some update here?
> 
> @QianNangong It would be better if you also share the original URL as well

https://piston-meta.mojang.com/v1/packages/76839fc33609dfd354b9c519f46bbea2afac9e06/1.19.json

**Comment by r58Playz** [comment]
This will most likely never happen.

**Comment by kb-1000** [comment]
@r58Playz don't advertise other launchers here.

Anyways, definitely not going to happen for your custom native hack.
However, since 1.19 Mojang ships ARM LWJGL. That _is_ planned. If you want that to happen sooner, consider to PR it.

As said in my first comment: This depends on Mojang shipping natives, which they now do. Now, it's just a matter of someone implementing it.

**Comment by 1oglop1** [comment]
@kb-1000 Is there any dev documentation or could you point me where to start so I do not have to start from zero?
I could invest some time into this.

**Comment by peterix** [comment]
Do not advertise your own launchers here.
